### PR TITLE
Sparse support for ReLU

### LINKED ISSUE #85208: Activation functions for sparse tensors.
### 🚀 The feature, motivation and pitch

Hello, 

in my model, I would like to apply a non-linear function such as the [ReLU](https://pytorch.org/docs/stable/generated/torch.nn.ReLU.html)  module or the [torch.nn.functional.relu](https://pytorch.org/docs/stable/generated/torch.nn.functional.relu.html) function to a [sparse](https://pytorch.org/docs/stable/sparse.html) tensor but get a ```NotImplementedError```. 
Could the common activation functions be implemented for sparse tensors?

### Alternatives

_No response_

### Additional context

_No response_

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86749

ReLU support for all sparse layouts, including backward.

Fixes #85208

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp [hunk 1]
```diff
@@ -384,6 +389,30 @@ Tensor& round_sparse_csr_(Tensor& self) {
...
 }
 
+Tensor threshold_backward_sparse_compressed(
+    const Tensor& grad_output,
+    const Tensor& self,
+    const Scalar& threshold) {
+  return get_result_tensor_for_unary_op(
+      [&](const Tensor& t) {
+        return at::threshold_backward(t, self.values(), threshold);
+      },
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:399]: Hmm, is there a guard that makes sure that `self` is coalesced? Otherwise `self.values()` will error out.

File: aten/src/ATen/native/sparse/SparseUnaryOps.cpp [hunk 2]
```diff
@@ -187,6 +192,28 @@ Tensor isinf_sparse_meta(const Tensor& self) {
...
+    return at::threshold_backward(t, self.values(), threshold);
+  });
+}
+
+Tensor& threshold_backward_sparse_out(
+    const Tensor& grad_output,
+    const Tensor& self,
+    const Scalar& threshold,
+    Tensor& grad_input) {
+  return coalesced_unary_ufunc_out(
```
Comment (amjames) [inline aten/src/ATen/native/sparse/SparseUnaryOps.cpp:211]: @nikitaved I think you meant for your comment to go here? Self will be coalesced as it should be stashed for the `grad_fn` immediately after return. At least that is what the tests would suggest. 

In truth I am not sure, do we need to consider the possibility that one might call this function on its own, as in not as the gradient of another function?

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/native/sparse/SparseUnaryOps.cpp (hunk 2 above)
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseUnaryOps.cpp:211]: `self` is used in forward. What if I pass an uncoalesced `self` into RELU?

Comment (amjames) [reply aten/src/ATen/native/sparse/SparseUnaryOps.cpp:211]: This is a bit confusing the but the `self` parameter here to the is actually the result from `ReLU` which will always be coalesced. I used the `self` name to be consistent with the strided implementation. The relevant declaration in derivatives.yaml:

```yaml
- name: relu(Tensor self) -> Tensor
  self: threshold_backward(grad, result, 0)
  result: auto_element_wise
```

Although, if we were to implement `threshold` then `self` would actually be `self` and your point holds we may attempt to access values without coalescence. I will update this.

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseUnaryOps.cpp:211]: Makes sense. Still `relu(uncoalesced input)` works just fine, right?

**Review by nikitaved — APPROVED** [review]
LGTM! Thank you, @amjames !
